### PR TITLE
Move blob to c10

### PR DESCRIPTION
**Pull request by smessmer**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#15853 Move blob to c10**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13605256/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15854 Code style cleanup&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13605258/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15855 Remove some dependencies from ivalue.h to ATen&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13605259/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15856 Fix export macros&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13605257/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15941 Split IValue -> Tensor dependency, move to c10&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13630156/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15954 Move c10 dispatcher to correct folder and namespace&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13633572/)

This is preparation work for moving IValue to c10.

Differential Revision: [D13605256](https://our.internmc.facebook.com/intern/diff/D13605256/)

### CODE REVIEW

File: c10/core/blob.h [hunk 1]
```diff
@@ -51,7 +49,7 @@ class CAFFE2_API Blob final : public c10::intrusive_ptr_target {
   /**
    * Returns the meta info of the blob.
    */
-  inline const TypeMeta& meta() const noexcept {
+  inline const caffe2::TypeMeta& meta() const noexcept {
```
Comment (ezyang) [inline c10/core/blob.h]: This looks weird

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
Can we get a comment saying that use of blob in c10 world is extremely not-preferred, because IValue is supposed to work for most of the use cases.

File: c10/core/blob.h (hunk 1 above)
Comment (smessmer) [reply c10/core/blob.h]: It is weird. `TypeMeta` is part of the c10 module and should be in the c10 namespace. The reason it still lives in the `caffe2` namespace is that the registration macros are called from the `caffe2` namespace and only work if the `TypeMeta` class is there as well. We'll have to fix this at some point.

**Comment by smessmer** [comment]
Yes, I'll add a comment. It's only meant to be used as a holder inside of `IValue`. Maybe I'll find an even better way, currently thinking about renaming it to `IValueAny`, moving it into the `detail` namespace, and have a `using caffe2::Blob = c10::detail::IValueAny` somewhere.

**Pull request closed without merging**